docs/src/API/configuration.md
```@meta
CurrentModule = Configuration(配置文件模块)
```

```@docs
#### mutable struct:
+ GENIE_VERSION(Genie版本-常量)
+ Settings(应用程序配置-设置应用程序的默认设置)
#### function:
+ isdev(判断是否为开发环境函数)
+ isprod(判断是否为生成环境函数)
+ istest(判断是否为测试环境函数)
+ env(获取当前环境字符串函数, 如: "dev")
+ buildpath(获取构建temp目录路径函数)
```


docs/src/guides/Interactive_environment.md
# Using Genie in an interactive environment (Jupyter/IJulia, REPL, etc)

Genie can be used for ad-hoc exploratory programming, to quickly whip up a web server
and expose your Julia functions.

Once you have `Genie` into scope, you can define a new `route`.
A `route` maps a URL to a function.

```julia
julia> import Genie.Router: route
julia> route("/") do
         "Hi there!"
       end
```

You can now start the web server using

```julia
julia> Genie.startup()
```

Finally, now navigate to <http://localhost:8000> – you should see the message "Hi there!".

We can define more complex URIs which can also map to previously defined functions:

```julia
julia> function hello_world()
         "Hello World!"
       end
julia> route("/hello/world", hello_world)
```

Obviously, the functions can be defined anywhere (in any other module) as long as they are accessible in the current scope.

You can now visit <http://localhost:8000/hello/world> in the browser.

Of course we can access GET params:

```julia
julia> import Genie.Router: @params
julia> route("/echo/:message") do
         @params(:message)
       end
```

Accessing <http://localhost:8000/echo/ciao> should echo "ciao".

And we can even match by types:

```julia
julia> route("/sum/:x::Int/:y::Int") do
         @params(:x) + @params(:y)
       end
```

By default, GET params are extracted as `SubString` (more exactly, `SubString{String}`).
If type constraints are added, Genie will attempt to convert the `SubString` to the indicated type.

For the above to work, we also need to tell Genie how to perform the conversion:

```julia
julia> import Base.convert
julia> convert(::Type{Int}, s::SubString{String}) = parse(Int, s)
```

Now if we access <http://localhost:8000/sum/2/3> we should see `5`

## Handling query string params

Query string params, which look like `...?foo=bar&baz=2` are automatically unpacked by Genie and placed into the `@params` collection. For example:

```julia
julia> route("/sum/:x::Int/:y::Int") do
         @params(:x) + @params(:y) + parse(Int, get(@params, :initial_value, "0"))
       end
```

Accessing <http://localhost:8000/sum/2/3?initial_value=10> will now output `15`.


CHANGELOG.md
# Changelog

## v1 - 2020-07-21

* added support for SSL to Genie apps
* added SSL initializer
* bundled localhost cert and key
* automatically forwarding all extra arguments from `Genie.up` to `HTTP.serve` `HTTP.listen`
* added tests
* fixed failing tests
* improved content negotiation and default error responses

## v0.32.0 - 2020-07-02

* new html() method, combining string view and file layout
* removed forgotten debug @show from Plugins
* fixed issue with <template> tags being stripped away
* fixed issue with missing <!DOCTYPE>
* deps update
* `Router.get_route` warns by default when named route not found (previously errored)

## v0.31.5 - 2020-07-01

* fixed issue with emtpy SECRET_TOKEN
* Session improvements
* more tests
* Markdown views fixes
* fixed `Genie.serve` for serving static websites

## v0.31.4 - 2020-07-01

* fixed issue with cookie encryption/decryption
* support for "inline" markdown views
* more tests

## v0.31.3 - 2020-06-30

* fix issue with CORS "*"
* documentation improvements
* doc comments

## v0.21.0 - 2020-04-23

* names of apps and resources are automatically cleaned (illegal Julia vars chars removed)
* support for file uploads with UTF-8 chars
* deps updates
* docs updates

## v0.25.2 - 2020-02-03

* bug fixes
* performance improvements
* deps update

## v0.22.9 - 2020-01-06

* refactoring of the Rendering engine (**breaking**)
* deps update

## v0.22.9 - 2019-12-24

* enhanced support for WebChannels and websockets operations

## v0.22.8 - 2019-12-05

* more tests
* support for `if` blocks
* internal API refactoring

## v0.22.7 - 2019-12-04

* added support for `if` blocks
* performance optimisations

## v0.22.6 - 2019-12-03

* bug fixes
* deps update

## v0.22.5 - 2019-11-30

* added type hints to renderer's `invokelastest` calls

## v0.22.4 - 2019-11-30

* fixed issue with JSON rendering where view variables won't get updated

## v0.22.3 - 2019-11-15

* logic to normalize the case of request headers to "Request-Header-Key"

## v0.21 - 2019-11-15

* deps update
* `Router.@params` is now set with default values
* vars in view can now be accessed directly (without `@vars`)
* internal API enhancements and bug fixes

## v0.20.0 - 2019-11-10

* internal rewrite and API cleanup (**breaking**)
* removal of Nullables and YAML deps
* deps update
* over 50 new unit tests added

## v0.19.1 - 2019-10-18

* Documentation improvements for `Genie.Deploy`
* Dockerfile generation enhancements
* Deps updates

## v0.19.0 - 2019-09-25

* New API for deployments
* Docker integration: generators for Dockerfile and API for building and running containers
* Heroku integration: API for deploying dockerized Genie apps to Heroku
* Extended API for app generation: `newapp_webservice`, `newapp_mvc`, and `newapp_fullstack`
* Deps updates

## v0.18.1 - 2019-09-25

* fixes an issue with parsing JSON payloads
* deps updates

## v0.18.0 - 2019-09-20

* pluralized some of the folders for consistent naming: sessions, tasks, and tests
* fixed an issue with generating a new task
* deps updates

## v0.17.1 - 2019-09-09

* improved support for testing
* deps updates
* small bug fixes

## v0.17.0 - 2019-08-31

* fully migrated logging to Julia's native logger (**breaking**)
* reintroduced support for logging to file via `LoggingExtras` (new app dependency)
* added a new command `bin/serverinteractive` allowing to start the web server interactively
* fixed issue with webserver port env settings being overwritten by default settings
* small changes to better support logging
* moved welcome info out of app files into core files
* added default `favicon.ico` file to avoid annoying 404 errors from automatic browsers requests
* cleaned up the env files
* small documentation updates
* dependencies updates

## v0.16.0 - 2019-08-29

* switched to native Julia logging (automatic logging to file for now is disabled, will come back in a future version)
* the `log` function has been removed (**breaking**)
* added support for embedded Julia within HTML arguments
* cleaned up HTML rendering
* refactored cache adapters loading to be less hacky and more performant
* refactored session adapters loading to be less hacky and more performant
* consolidated the `Helpers` API into the `Requests` and `Sessions` modules and removed `Helpers` module (**breaking**)
* added new `Exceptions` module defining the `ExceptionalResponse` type
* added extra `@params` pointing to the currently matched route and webchannel
* fixed broken `Cookies` and `Session` functionality
* `Renderer.redirect` now supports extra arguments which are passed to `Router.linkto` for reverse routing
* new `Renderer.response` method specialized for `ExceptionalResponse`
* consolidated `flash` functionality in dedicated module `Flash` (**breaking**)
* added support for URI segments matching in routes
* refactored the `Route` and `Channel` types
* `ExceptionalResponses` now break the execution flow if thrown from controller hooks
* added `up()` as shortcut for `Genie.startup()` to start the web servers
* internal API consolidation
* added new generic `error-xxx.html` page template
* updated bundled JS and CSS files to newer versions

## v0.15.0 - 2019-08-22

* fixed error in `Genie.newapp()` with `dbsupport = true`
* internal API cleanup and optimisations
* fixed issue with `newresource` SearchLight integration
* SearchLight initializer code is now uncommented
* dependencies update
* `Router.tolink` and its alias `Router.linkto` throw exceptions if the route is not defined (**breaking**)
* `Router.tolink!!` and its alias `Router.linkto!!` have been removed (**breaking**)
* new method `Requests.read(HttpFile, Type{String})` which returns the content of an uploaded file as a string.

## v0.14.0 - 2019-08-21

* consolidation of the Generator API

src/Configuration.jl
"""
Core genie configuration / settings functionality.
核心精灵配置/设置功能。
"""
module Configuration

using Revise

"""
  const GENIE_VERSION

References the current Genie version number.
引用当前的Genie版本号。
"""
const GENIE_VERSION = v"1.1"

import Logging
import Genie
# 加密相关包
import MbedTLS

export isdev, isprod, istest, env
export Settings, DEV, PROD, TEST

# app environments
const DEV   = "dev"
const PROD  = "prod"
const TEST  = "test"

# 不存在时 设置环境默认值
haskey(ENV, "GENIE_ENV") || (ENV["GENIE_ENV"] = DEV)
haskey(ENV, "HOST") || (ENV["HOST"] = "127.0.0.1")


"""
    isdev()  :: Bool

Set of utility functions that return whether or not the current environment is development, production or testing.
一组实用函数，返回当前环境是开发、生产还是测试。
# Examples
```julia
julia> Configuration.isdev()
true

julia> Configuration.isprod()
false
```
"""
isdev() :: Bool  = (Genie.config.app_env == DEV)


"""
    isprod() :: Bool

Set of utility functions that return whether or not the current environment is development, production or testing.
一组实用函数，返回当前环境是开发、生产还是测试。
# Examples
```julia
julia> Configuration.isdev()
true

julia> Configuration.isprod()
false
```
"""
isprod():: Bool = (Genie.config.app_env == PROD)


"""
    istest() :: Bool

Set of utility functions that return whether or not the current environment is development, production or testing.
一组实用函数，返回当前环境是开发、生产还是测试。

# Examples
```julia
julia> Configuration.isdev()
true

julia> Configuration.isprod()
false
```
"""
istest():: Bool = (Genie.config.app_env == TEST)


"""
    env() :: String

Returns the current Genie environment.
返回当前的精灵环境。
# Examples
```julia
julia> Configuration.env()
"dev"
```
"""
env() :: String = Genie.config.app_env


"""
    buildpath() :: String

Constructs the temp dir where Genie's view files are built.
构造构建精灵视图文件的temp目录。
"""
buildpath() :: String = Base.Filesystem.mktempdir(prefix = "jl_genie_build_", cleanup = false)


"""
    mutable struct Settings

App configuration - sets up the app's defaults. Individual options are overwritten in the corresponding environment file.
应用程序配置-设置应用程序的默认设置。在相应的环境文件中覆盖各个选项。

# Arguments
- `server_port::Int`: the port for running the web server (default 8000)
                      用于运行web服务器的端口(默认为8000)
- `server_host::String`: the host for running the web server (default "127.0.0.1")
                         运行web服务器的主机(默认为“127.0.0.1”)
- `server_document_root::String`: path to the document root (default "public/")
                                  文档根目录的路径(默认为“public/”)
- `server_handle_static_files::Bool`: if `true`, Genie will also serve static files. In production, it is recommended to serve static files with a web server like Nginx.
                                      如果“true”，Genie还将提供静态文件。在生产中，建议使用像Nginx这样的web服务器来提供静态文件。
- `server_signature::String`: Genie's signature used for tagging the HTTP responses. If empty, it will not be added.
                              Genie用于标记HTTP响应的签名。如果为空，则不会添加。
- `app_env::String`: the environment in which the app is running (dev, test, or prod)
                     应用程序运行的环境(dev、test或prod)
- `cors_headers::Dict{String,String}`: default `Access-Control-*` CORS settings
                                       默认的“Access-Control-*” CORS设置
- `cors_allowed_origins::Vector{String}`: allowed origin hosts for CORS settings
                                          允许原点主机进行CORS设置
- `cache_duraction::Int`: cache expiration time in seconds
                          缓存过期时间(秒)
- `log_level::Logging.LogLevel`: logging severity level
                                 日志严重性级别
- `log_to_file::Bool`: if true, information will be logged to file besides REPL
                       如果为真，除REPL外，信息将被记录到文件中
- `assets_fingerprinted::Bool`: if true, asset fingerprinting is used in the asset pipeline
                                如果为真，则在资产管道中使用资产指纹(认证好理解点)
- `session_key_name::String`: the name of the session cookie
                              会话cookie的名称
- `session_storage::Symbol`: the backend adapter for session storage (default File)
                             会话存储的后端适配器(默认文件)
- `inflector_irregulars::Vector{Tuple{String,String}}`: additional irregular singular-plural forms to be used by the Inflector
                                                        变形器将使用其他不规则单数复数形式
- `run_as_server::Bool`: when true the server thread is launched synchronously to avoid that the script exits
                         如果为true，则同步启动服务器线程，以避免脚本退出
- `websockets_server::Bool`: if true, the websocket server is also started together with the web server
                             如果为true，则还将与Web服务器一起启动websocket服务器
- `html_parser_close_tag::String`: default " /". Can be changed to an empty string "" so the single tags would not be closed.
                                   默认为" /"。 可以更改为空字符串""，这样就不会关闭单个标签。
- `ssl_enabled::Bool`: default false. Server runs over SSL/HTTPS in development.
                       默认为false。 服务器在开发中通过SSL / HTTPS运行。
- `ssl_config::MbedTLS.SSLConfig`: default `nothing`. If not `nothing` and `ssl_enabled`, it will use the config to start the server over HTTPS.
                                   默认为"nothing"。 如果不是nothing和ssl_enabled，它将使用配置通过HTTPS启动服务器
"""
mutable struct Settings
  # 服务器端口
  server_port::Int
  #服务器ip
  server_host::String
  # 服务器根目录
  server_document_root::String
  # 处理服务器静态文件
  server_handle_static_files::Bool
  # 服务器签名
  server_signature::String
  # 应用环境
  app_env::String
  # 基站头信息
  cors_headers::Dict{String,String}
  # 基站允许源
  cors_allowed_origins::Vector{String}


  # 缓存时间
  cache_duration::Int
  # 缓存存储
  cache_storage::Symbol

  # 日志等级
  log_level::Logging.LogLevel
  # 日志是否保存到文件
  log_to_file::Bool

  # 资产(?，资源)-鉴定
  assets_fingerprinted::Bool

  # 会话 key 名称
  session_key_name::String
  # 会话存储
  session_storage::Symbol

  # 传播器-不定时
  inflector_irregulars::Vector{Tuple{String,String}}

  # 作为服务器运行
  run_as_server::Bool

  # 是否作为websocket服务器
  websockets_server::Bool
  # websocket端口
  websockets_port::Int

  # 设定项文件夹初始值
  initializers_folder::String

  # 路径-配置
  path_config::String
  # 路径-环境
  path_env::String
  # 路径-应用
  path_app::String
  # 路径-资源
  path_resources::String
  # 路径-库文件
  path_lib::String
  # 路径-助手
  path_helpers::String
  # 路径-日志
  path_log::String
  # 路径-任务
  path_tasks::String
  # 路径-构建
  path_build::String
  # 路径-插件
  path_plugins::String
  # 路径-缓存
  path_cache::String
  # 路径-构造器
  path_initializers::String
  # 路径-数据库
  path_db::String
  # 路径-二级制文件
  path_bin::String
  # 路径-源
  path_src::String

  # web管道-默认路由
  webchannels_default_route::String
  # web管道-js文件
  webchannels_js_file::String
  # web管道-订阅频道(?)
  webchannels_subscribe_channel::String
  # web管道-取消订阅频道(?)
  webchannels_unsubscribe_channel::String
  # web管道-默认路由
  webchannels_autosubscribe::Bool

  # html解析器-关闭标记
  html_parser_close_tag::String
  # html解析器-字符在(?)
  html_parser_char_at::String
  # html解析器-字符点(?)
  html_parser_char_dot::String
  # html解析器-字符列(?)
  html_parser_char_column::String
  # html解析器-字符破折符
  html_parser_char_dash::String

  # ssl证书启动
  ssl_enabled::Bool
  # ssl证书配置
  ssl_config::Union{MbedTLS.SSLConfig,Nothing}

  # 默认构造
  Settings(;
            server_port                 = (haskey(ENV, "PORT") ? parse(Int, ENV["PORT"]) : 8000), # default port for binding the web server
            server_host                 = ENV["HOST"],
            server_document_root        = "public",
            server_handle_static_files  = true,
            server_signature            = "Genie/$GENIE_VERSION/Julia/$VERSION",

            app_env                     = ENV["GENIE_ENV"],

            cors_headers  = Dict{String,String}(
              "Access-Control-Allow-Origin"       => "", # ex: "*" or "http://mozilla.org"
              "Access-Control-Expose-Headers"     => "", # ex: "X-My-Custom-Header, X-Another-Custom-Header"
              "Access-Control-Max-Age"            => "86400", # 24 hours
              "Access-Control-Allow-Credentials"  => "", # "true" or "false"
              "Access-Control-Allow-Methods"      => "", # ex: "POST, GET"
              "Access-Control-Allow-Headers"      => "", # ex: "X-PINGOTHER, Content-Type"
            ),
            cors_allowed_origins = String[],

            cache_duration    = 0,
            cache_storage     = :File,

            log_level     = Logging.Debug,
            log_to_file   = false,

            assets_fingerprinted  = false,

            session_key_name    = "__geniesid",
            session_storage     = :File,

            inflector_irregulars = Tuple{String,String}[],

            run_as_server = false,

            websockets_server = false,
            websockets_port   = server_port + 1,

            initializers_folder = "initializers",

            path_config         = "config",
            path_env            = joinpath(path_config, "env"),
            path_app            = "app",
            path_resources      = joinpath(path_app, "resources"),
            path_lib            = "lib",
            path_helpers        = joinpath(path_app, "helpers"),
            path_log            = "log",
            path_tasks          = "tasks",
            path_build          = buildpath(),
            path_plugins        = "plugins",
            path_cache          = "cache",
            path_initializers   = joinpath(path_config, initializers_folder),
            path_db             = "db",
            path_bin            = "bin",
            path_src            = "src",

            webchannels_default_route       = "__",
            webchannels_js_file             = "channels.js",
            webchannels_subscribe_channel   = "subscribe",
            webchannels_unsubscribe_channel = "unsubscribe",
            webchannels_autosubscribe       = true,

            html_parser_close_tag = " /",
            html_parser_char_at = "!!",
            html_parser_char_dot = "!",
            html_parser_char_column = "!",
            html_parser_char_dash = "__",

            ssl_enabled = false,
            ssl_config = nothing
        ) =
              new(
                  server_port, server_host,
                  server_document_root, server_handle_static_files, server_signature,
                  app_env,
                  cors_headers, cors_allowed_origins,
                  cache_duration, cache_storage,
                  log_level, log_to_file,
                  assets_fingerprinted,
                  session_key_name, session_storage,
                  inflector_irregulars,
                  run_as_server,
                  websockets_server, websockets_port,
                  initializers_folder,
                  path_config, path_env, path_app, path_resources, path_lib, path_helpers, path_log, path_tasks, path_build,
                  path_plugins, path_cache, path_initializers, path_db, path_bin, path_src,
                  webchannels_default_route, webchannels_js_file, webchannels_subscribe_channel, webchannels_unsubscribe_channel, webchannels_autosubscribe,
                  html_parser_close_tag, html_parser_char_at, html_parser_char_dot, html_parser_char_column, html_parser_char_dash,
                  ssl_enabled, ssl_config
                )
end

end
